include/tvm/relay/expr_visitor.h

```diff
@@ -0,0 +1,186 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file tvm/relay/expr_visitor.h
+ * \brief A simple visitor wrapper around ExprFunctor.
+ *
+ * Exposes two visitors with default traversal strategies, one
+ * which doesn't compute a result but can mutate internal state,
+ * and another which functionally builds a new Expr.
+ */
+#ifndef TVM_RELAY_EXPR_VISITOR_H_
+#define TVM_RELAY_EXPR_VISITOR_H_
+
+#include <tvm/relay/expr_functor.h>
+
+namespace tvm {
+namespace relay {
+
+class ExprVisitor : public ::tvm::relay::ExprFunctor<void(const Expr& n)> {
+ public:
+  void VisitExpr_(const VarNode* op) override { return; }
+
+  void VisitExpr_(const GlobalVarNode* op) override { return; }
+
+  void VisitExpr_(const ConstantNode* op) override { return; }
+
+  void VisitExpr_(const TupleNode* op) override {
+    for (auto field : op->fields) {
+      this->VisitExpr(field);
+    }
+  }
+
+  void VisitExpr_(const ParamNode* op) override { this->VisitExpr(op->var); }
+
+  void VisitExpr_(const FunctionNode* op) override {
+    for (auto param : op->params) {
+      this->VisitExpr(param);
+    }
+
+    this->VisitExpr(op->body);
+  }
+
+  void VisitExpr_(const CallNode* op) override {
+    this->VisitExpr(op->op);
+    for (auto ty_arg : op->type_args) {
+      this->VisitType(ty_arg);
+    }
+
+    for (auto arg : op->args) {
+      this->VisitExpr(arg);
+    }
+  }
+
+  void VisitExpr_(const LetNode* op) override {
+    this->VisitExpr(op->var);
+    this->VisitExpr(op->value);
+    this->VisitExpr(op->body);
+  }
+
+  void VisitExpr_(const IfNode* op) override {
+    this->VisitExpr(op->cond);
+    this->VisitExpr(op->true_branch);
+    this->VisitExpr(op->false_branch);
+  }
+
+  void VisitExpr_(const OpNode* op) override { return; }
+
+  virtual void VisitType(const Type& t) {}
+};
+
+class ExprFVisitor : public ::tvm::relay::ExprFunctor<Expr(const Expr& n)> {
```

Comment:
There are two things that we need to improve the Mutator:

- Return the expr itself when there is no change in the children
- Memoize the result, so if we call Muate on the same expr, it will return memoized result